Translate all job statuses to Qiskit terminology in client

caleb-johnson opened the pull request:
Fixes #1039 

### Summary
In this PR, we translate job statuses from all job clients to Qiskit terminology. Rather than changing the DB structure (currently uses ray status names), we will just do a translation in the serverless client when the user requests the status.


**Serverless --> Qiskit nomenclature changes**

Pending --> Initializing
Stopped --> Canceled
Succeeded --> Done
Failed --> Error

IceKhan13 commented on gateway/api/migrations/0001_initial.py:
```diff
@@ -94,14 +94,14 @@ class Migration(migrations.Migration):
                     models.CharField(
                         choices=[
-                            ("PENDING", "Pending"),
+                            ("INITIALIZING", "Initializing"),
```
Migrations should not be changed. They are append only.
Since they are changing structure of schema of database, they can only modify state of db that was there before. 

You need to create new migration that will change schema from this types of statuses to new ones.

caleb-johnson replied:
I think handling the migration on client side, like you suggested, is a nice solution here. I don't even think it's really that hacky. We leave everything on the backend in Ray language for clarity, and we just convert to our terminology on client side. Makes a lot of sense to me.

I can also generate a new migration file if you like that solution better.

IceKhan13 reviewed (changes requested):
Also, when changing records in existing db you need to take care of entries in DB that are already there and convert them too. 

---
Other hotfix options would be just to do mapping in client, so you will not need to change anything on a backend and just show in client mapped statuses. Something like
```python
def gateway_to_runtime_statuses(gateway_status):
     mapping = {
         "SUCCEEDED": "DONE"
     }
     return mapping.get(gateway_status, f"APIStatus[{gateway_status}]")
```

IceKhan13 commented:
Basically what migrations are doing:
- they are checking in which state db is (which migration version is applied to DB recenty)
- if there are upapplied migrations, they are applied
- if any of previous migrations changed, migration will fail

caleb-johnson commented on client/quantum_serverless/core/job.py:
```diff
@@ -127,7 +127,7 @@ def __init__(self, client: JobSubmissionClient):
     def status(self, job_id: str):
-        return self._job_client.get_job_status(job_id)
+        return self._job_client.get_job_status(job_id).value
 
```
Since `status` is an unimplemented `BaseJobClient` method, I think both the ray and gateway job clients should return a string indicating the status. I updated this in this PR because it affects how I do the status translation in the user-facing job client

IceKhan13 replied:
so, you basically do not want to return status class? and return status string?
I have no objections, just to confirm :)

caleb-johnson replied:
I figured since the gateway job client returns a string, we may as well return a string from the ray job client as well, since they both derive from the same base class. Not a big deal, but I thought it was a clean simplification.

caleb-johnson commented on client/quantum_serverless/core/job.py:
```diff
@@ -473,7 +473,7 @@ def __init__(
     def status(self):
         """Returns status of the job."""
-        return self._job_client.status(self.job_id)
+        return _map_status_to_serverless(self._job_client.status(self.job_id))
```
We will do the translation from any supported job client status to Qiskit terminology here in the user-facing `Job` client. This is the only place we'll do the translation. Everything underneath this will be Ray terminology.

caleb-johnson commented on client/quantum_serverless/core/job.py:
```diff
@@ -496,7 +496,7 @@ def result(self, wait=True, cadence=5, verbose=False):
             if verbose:
                 logging.info("Waiting for job result.")
-            while not self._in_terminal_state():
+            while not self.in_terminal_state():
```
We don't have a `JobStatus` object like ray that we can query like Ray's `JobStatus.is_terminal`, so we will make this `Job.in_terminal_state` method public, so clients have a way to query whether the job has finished.

caleb-johnson commented:
@IceKhan13 , I've fixed this up as you suggested. I don't currently have an idea about this strange sphinx error, but Im looking into it. Doesn't seem to be related to the work in this PR, so it's a little confusing

caleb-johnson commented:
> @IceKhan13 , I've fixed this up as you suggested. I don't currently have an idea about this strange sphinx error, but Im looking into it. Doesn't seem to be related to the work in this PR, so it's a little confusing

Hmm, I fixed it by removing `Program` from the init file, but it is still not clear to me why that error occured in this PR.

caleb-johnson commented on client/quantum_serverless/core/__init__.py:
```diff
@@ -39,7 +39,6 @@
     RayJobClient
     save_result
     QiskitPattern
-    Program
```
This is causing Sphinx errors for me since it is no longer imported in this file, not sure how main branch is passing sphinx build with no warnings?

caleb-johnson commented on client/quantum_serverless/core/job.py:
```diff
@@ -723,3 +723,20 @@ def save_result(result: Dict[str, Any]):
+    try:
+        return status_map[status]
+    except KeyError as exc:
+        raise KeyError(f"Cannot interpret unknown job status: {status}") from exc
```
Maybe we shouldn't error, maybe we should just print the status like you suggested? Not really sure

IceKhan13 replied:
I think printing is less invasive :)

IceKhan13 commented:
> Hmm, I fixed it locally by removing Program from the docstring in the init file, but the error is still happening in CI. Can't reproduce this locally after removing Program from the docstring.

merge main and you will be good :)

IceKhan13 commented:
PR looks good! simple and clean like all great things 😃

IceKhan13 reviewed (approved):
LGTM
